Analytics Performance: Support custom selection timestamps

ThomazFB opened the pull request:
Summary
==========
Fix issue #9269 by introducing the same timed refresh constraint to Custom ranges selection inside the Hub. 

How to Test
==========
1. Open the app and enter the Analytics Hub
2. Once the data is loaded, select a Custom range selection, remember the one you selected and wait the loading to finish.
3. Once your custom range selection loading ends, leave the Hub, go back again and select the same Custom range you selected before, verify that now the cached data is used instead of a full triggering a full load.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Update AnalyticsUpdateDataStore to fully support CUSTOM selection type timestamp operations

ThomazFB committed:
Remove CUSTOM constraint from UpdateAnalyticsHubStats

ThomazFB committed:
Update CUSTOM test scenario to make sure selection is not distinguished anymore

ThomazFB committed:
Fix UpdateAnalyticsHubStatsTest failing test

ThomazFB committed:
Merge branch 'issue/9269-adjust-analytics-hub-to-use-stored-data' into issue/9269-support-custom-selection-timestamps

ThomazFB committed:
Refactor shouldUpdateAnalytics operation

ThomazFB committed:
Further refactor shouldUpdateAnalytics expressions

ThomazFB committed:
Merge branch 'issue/9269-adjust-analytics-hub-to-use-stored-data' into issue/9269-support-custom-selection-timestamps

atorresveiga commented:
Could it be that something broke with the selection?

I cannot change the selection on the Analytics Hub screen, I tried with different options and even with the Custom but the Today option remains selected:

https://github.com/woocommerce/woocommerce-android/assets/18119390/7909e981-6ae8-4b11-a0c0-b6938d5415b5

ThomazFB commented:
Hey @atorresveiga, nasty bug you manage to find, and I'm glad we found it in time! The https://github.com/woocommerce/woocommerce-android/pull/9320 PR introduced it, and the culprit is the Flow we get from the new DataStore that makes the `UpdateAnalyticsHubStats` stuck and unresponsive. I'm working on a solution that can fit with the DataStore API that works exclusively with Flow to see if we can get this fixed reliably! Thanks!

ThomazFB committed:
Merge branch 'issue/9269-adjust-analytics-hub-to-use-stored-data' into issue/9269-support-custom-selection-timestamps

ThomazFB commented:
@atorresveiga, the issue you reported is now fixed through the PR https://github.com/woocommerce/woocommerce-android/pull/9320, and I updated this one here with the code there. This PR is ready for a second round! Thanks!

You can check the fix through these three commits: https://github.com/woocommerce/woocommerce-android/pull/9320/commits/7aa4dd907763c5033a51eda293108326c420d432, https://github.com/woocommerce/woocommerce-android/pull/9320/commits/52dfc8fc8926479b57693d31943f7966f2d88679 and https://github.com/woocommerce/woocommerce-android/pull/9320/commits/e51294d44384c3a537fa9fff48d71e7533ee814f

ThomazFB committed:
Merge branch 'issue/9269-adjust-analytics-hub-to-use-stored-data' into issue/9269-support-custom-selection-timestamps

atorresveiga commented:
Great work Thomaz! The selection is working again 😄.
~~I tested the changes but the cached data is used only the third time I use the same custom selection. I recorded a video:~~

I didn't check that the whole screen finishes loading before making the second attempt 🤦 

https://github.com/woocommerce/woocommerce-android/assets/18119390/364a6871-9988-4f56-8e63-decb7aaec25c

atorresveiga reviewed (approved):
The code looks nice and is working as expected! :shipit:
